### PR TITLE
Combine Sharepoint and OneDrive - Code changes

### LINKED ISSUE #263: OneDrive 2.2 set to 30 days or less
# 💡 Summary #

Change from: "“Expiration timers SHOULD be set to 30 days.” to "“Expiration timers SHOULD be set to 30 days or less."

## Motivation and context ##

Currently 2.2 is set to 30 days only but expiration times less than 30 days should also be accepted 

## Implementation notes ##

Please provide details for implementation, such as:

- change the wording to accept expiration time that is less than 30 days

## Acceptance criteria ##

How do we know when this work is done?

- [x] updated policy has the changes noted in Implementation notes.

### LINKED ISSUE #264: Split OneDrive 2.3 into two policies
# 💡 Summary #

Split and clarify the current OneDrive 2.3 baseline v.0.1 policy "Link Permissions SHOULD Be Set to Enabled Anyone Links to View" into two separate policies to make it easier for readers to understand and for automation to report.

## Motivation and context ##

The current baseline policy is a solid best practice, but the wording does not make it clear that there are two separate settings needed to properly protect links: one to make the default sharing link only be to others within the organization (i.e., not anyone) and a second to make sure when a user does make a link available to anyone that the default permissions on that link are view only.
This would be useful because right now the baseline does reference the Microsoft best practice that includes both settings, but the implementation section 2.3.4 in the baseline itself only shows the setting changes to address the view only portion.

## Implementation notes ##

Recommended update would be to split this policy into two policies to address both the default link sharing type and the anyone link default permissions. Both policies would still include references to the same Microsoft best practices document, but the implementation details would differ.
The new policies would be something like the following:

2.3 Default Link Sharing Type SHOULD NOT Be Set to Anyone
If Anyone link sharing is enabled, the default when creating a link is set to Anyone. Providing a more secure default option can mitigate unintentional sharing with unauthenticated anonymous third parties. To do so, set the default sharing type to be limited to people within your organization instead.
The implementation details would then be as follows:
To set the default file and folder sharing link for the organization:
1.	Open the SharePoint admin center, and select Sharing.
2.	Under File and folder links, select Only people in your organization or Specific people (only the people the user specifies).
3.	Select Save

2.4 Anyone link permissions SHOULD be Set to View Only
The implementation section would include instructions
1.	Open the SharePoint admin center.
2.	In the left-hand navigation pane, expand Policies, then select Sharing.
3.	Under Advanced settings for Anyone links, set the Files permissions to View.
5.	Under Advanced settings for Anyone links, set the Folders permissions to View.

## Acceptance criteria ##

How do we know when this work is done?

Updated policy has the changes noted in Implementation notes.

### LINKED ISSUE #391: OneDrive Code change to reflect the baseline changes
# 💡 Summary #

OneDrive Code change to reflect the baseline changes

## Motivation and context ##

baseline updated/changed for OneDrive, code should reflect those changes

## Implementation notes ##

- Rego code Changes 
- Powershell Code
- Unit test Code

## Acceptance criteria ##

How do we know when this work is done?

- [ ] code changes for rego
- [ ] code changes for powershell
- [ ] code changes for unit test code

### PR DESCRIPTION
**Pull request by Dylan-MITRE**
## 🗣 Description ##

Update the OneDrive / Sharepoint ScubaGear Code to reflect the merge of the two baselines and updates to policies.

## 💭 Motivation and context ##

Changes overview:
Provider code: remove old providers
Provider cove: create/edited Sharepoint to combine both sharepoint and onedrive
Rego code: remove old providers
Rego cove: create/edited Sharepoint to combine both sharepoint and onedrive
Unit tests: removed old unit tests
Unit tests: create/edited Sharepoint to combine both sharepoint and onedrive
Reporter module/Orchestrator/Connection: removed onedrive related lines 
README

Closes #391 
Closes #264 
Closes #263 

## 🧪 Testing ##

Run Scuba against Sharepoint 

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

### CODE REVIEW

**Comment by tkol2022** [comment]
### Question on missing powershell module
Why was the MS Online Sharepoint Powershell removed from the dependencies?

image

**Comment by tkol2022** [comment]
### Report is empty
I tried executing the Sharepoint product with the updated code and received a blank report output in two different test tenants. No Powershell errors.

image

**Comment by Dylan-MITRE** [comment]
> ### Question on missing powershell module
> Why was the MS Online Sharepoint Powershell removed from the dependencies?
> 
> image

must accidentally removed the requirement. i will add it back

**Comment by Dylan-MITRE** [comment]
> ### Report is empty
> I tried executing the Sharepoint product with the updated code and received a blank report output in two different test tenants. No Powershell errors.
> 
> image

its due to the baseline file's name was shrepointonedrive, i have changed the baseline file name, it should work now

File: Rego/SharepointConfig.rego [hunk 1]
```diff
@@ -110,88 +143,143 @@ tests[{
 }
 #--
 
-###################
-# MS.SHAREPOINT.3 #
-###################
-
 #
-# MS.SHAREPOINT.3.1v1
+# MS.SHAREPOINT.2.2v1
```
Comment (Sloane4) [inline Rego/SharepointConfig.rego:147]: Comment structure is wrong, should be:
```
#
# MS.SHAREPOINT.2.2v1
#--

# SPO_tenant - DefaultLinkPermission
# 1 view 2 edit
```

File: Rego/SharepointConfig.rego [hunk 2]
```diff
@@ -53,10 +84,11 @@ tests[{
 #--
 
 #
-# MS.SHAREPOINT.1.3v1
+# MS.SHAREPOINT.1.4v1
 #--
```
Comment (Sloane4) [inline Rego/SharepointConfig.rego:88]: Should be a newline after this line to break up declaration of start of new policy & comment notes for code

File: Rego/SharepointConfig.rego [hunk 3]
```diff
@@ -240,14 +328,13 @@ tests[{
...
 
 ###################
-# MS.SHAREPOINT.5 #
+# MS.SHAREPOINT.4 #
 ###################
 
 #
-# MS.SHAREPOINT.5.1v1
+# MS.SHAREPOINT.4.1v1
 #--
```
Comment (Sloane4) [inline Rego/SharepointConfig.rego:338]: Should be a newline after this line to break up declaration of start of new policy & comment notes for code

File: Testing/Unit/Rego/Sharepoint/SharepointConfig_01_test.rego [hunk 4]
```diff
@@ -82,9 +82,90 @@ test_SharingCapability_Incorrect_V2 if {
...
+        "OneDrive_PnP_Flag": false   
+    }
+
+    RuleOutput := [Result | Result = Output[_]; Result.PolicyId == PolicyId]
+
+    count(RuleOutput) == 1
+    not RuleOutput[0].RequirementMet
+    RuleOutput[0].ReportDetails == "Requirement not met"
+}
+#--
```
Comment (Sloane4) [inline Testing/Unit/Rego/Sharepoint/SharepointConfig_01_test.rego:161]: I do not see any test cases with `"OneDrive_PnP_Flag": true`

File: Testing/Unit/Rego/Sharepoint/SharepointConfig_03_test.rego [hunk 5]
```diff
@@ -1,20 +1,324 @@
...
+            }
+        ]
+    }
+
+    RuleOutput := [Result | Result = Output[_]; Result.PolicyId == PolicyId]
+
+    count(RuleOutput) == 1
+    not RuleOutput[0].RequirementMet
+    RuleOutput[0].ReportDetails == "Requirement not met: files are not limited to view for Anyone"
+}
```
Comment (Sloane4) [inline Testing/Unit/Rego/Sharepoint/SharepointConfig_03_test.rego:200]: I do not see a check for the OneDrive_PnP_Flag

**Review by tkol2022 — APPROVED** [review]

**Comment by tkol2022** [comment]
### Test Summary
Performed some basic stability tests and ensured that the product functions without errors. Detailed functional testing of the policies will be performed by the Automated Functional Test orchestrator which Crutch has developed - those tests may reveal logic bugs which will be opened as separate GitHub issues.

**Comment by Dylan-MITRE** [comment]
> Received errors in output file: image image

oh that's related to the powershell script. my recent update should resolved that

**Review by ssatyapal123 — APPROVED** [review]
Reviewed fixes and the errors are resolved. Though not part of this PR, the README.md should reflect the argument change from "spo" to "sharepoint".

**Comment by Dylan-MITRE** [comment]
> Reviewed fixes and the errors are resolved. Though not part of this PR, the README.md should reflect the argument change from "spo" to "sharepoint".

changes made

**Review by Sloane4 — APPROVED** [review]

**Comment by Sloane4** [comment]
@Dylan-MITRE don't forget to update all the tests

**Comment by Dylan-MITRE** [comment]
@nanda-katikaneni should be ready to merge (not sure the powershell tests are related to this)

**Pull request merged**